app/controllers/administrate/application_controller.rb

```diff
@@ -283,6 +300,19 @@ def authorize_resource(resource)
       end
     end
 
+    def authorized_new_resource
+      new_resource.tap do |resource|
+        authorize_resource(resource)
+      end
+    end
+    helper_method :authorized_new_resource
+
+    # Override this if you want to contextualize the resource differently.
+    #
+    # @param resource [ActiveRecord::Base] A resource to be contextualized.
+    # @return nothing
+    def contextualize_resource(resource); end
```

Comment:
I'm not sure that I understand this method. It's intended to make changes on the resource, but the context seems better suited to live in the `Page` object (which has access to the resource anyway). I'm not sure I would want to alter the model in this way.

Would you be able to explain this a bit more? 🙂

Reply:
Thank you for your review!! I have been waiting for your response for a long time.

I would like to contextualize the `Page` object, but I also intend to contextualize the resource in the Controller. By contextualizing the resource, we can use it to branch validations or trigger callbacks.
I believe that `contextualize_resource` should be separate from the `Page` 's context, as it serves a similar function to `scoped_resources` and `authorize_resource` .

Reply:
OK, I think I see now what is going on 👍 

When contextualizing, we can hide fields and then put default values on them. In your example you do this to use the `customer` field from non-admins, then set its value to the current user on new/create.

I'm not convinced about the following:

- Doing this as part of `authorize_resource`, which is a separate concern.
- Doing this only when `resource.is_a?(ActiveRecord::Base)`, which seems artificial. OK, it's not like we have currently support for anything other than ActiveRecord, but this explicit check seems strange to me.
- Having this new API instead of the existing `new_resource` and `requested_resource` instead. It's two separate steps, but it uses existing, known APIs. I have tried locally and it appears to work (I'm having issues running specs so I can't be 100% sure).

Reply:
Thank you for your review.

I have separated `contextize_resource` from `authorize_resource`. As a result, the resource no longer passes as a model class from the index, which also resolves the `ActiveRecord::Base` issue.

I believe the current form of `requested_resource` is ideal. What do you think? I would like to introduce the same form for `new` and `create`. Do you have any good suggestions on how to do this?

For example, what do you think about this form?

```ruby
def built_resource(params: {})
  @built_resource = new_resource(params).tap do |resource|
    authorize_resource(resource)
    contextualize_resource(resource)
  end
end
```

Reply:
Shifting slightly to a different topic, I feel that the current action methods in the Administrate's ApplicationController are quite fat, which I believe limits the customizability for us, the application developers.
Therefore, it would be beneficial if these methods were broken down further, allowing us to override specific methods during application development.

I believe it would be ideal if we could proceed with method division and achieve a more unified form as shown below.
(It might be beneficial to also divide `page = Administrate::Page::Form.new` and the part where context is injected into `Page` into separate methods.)

```ruby
    def new
      page = Administrate::Page::Form.new(dashboard, built_resource)
      render ...

    def edit
      page = Administrate::Page::Form.new(dashboard, requested_resource)
      render ...

    def create
      if built_resource(resource_params).save
      ...

    def update
      if requested_resource.update(resource_params)
      ...

    def built_resource(params: {})
      @built_resource ||= new_resource(params).tap do |resource|
        authorize_resource(resource)
        contextualize_resource(resource)
      end
    end

    def requested_resource
      @requested_resource ||= find_resource(params[:id]).tap do |resource|
        authorize_resource(resource)
        contextualize_resource(resource)
      end
    end
```

 What do you think?

Reply:
I think the other two commits are good ideas. Thank you.

Reply:
The reason I prepared built_resource in addition to new_resource is because I thought it might affect the following process that is used as a helper_method from the view. Do you foresee any issues with this?
https://github.com/search?q=repo%3Athoughtbot%2Fadministrate+new_resource+language%3AHTML%2BERB&type=code&l=HTML%2BERB